### PR TITLE
SAK-44945 BasicLTI Add the ability to assign more than one site ID to a single LTI tool

### PR DESCRIPTION
**Pull request by zhuoY121**
https://sakaiproject.atlassian.net/browse/SAK-44945

1. Create a table called `lti_tool_site` to store information about the visibility of tools on different sites. 
- This table only indicates visibility and does not grant editing powers to instructors, meaning deploying a tool to a site does not confer editing rights.

2. Create web pages that allow users to perform CRUD operations on the `lti_tool_site` table.

3. Add a new parameter `includeLaunchable` to the `getThingsDao()` method. This acts as a switch to enable or disable the feature. When set to true, non-Admins can see tools deployed to a site.

4. Now, tools are launchable after being placed in all the placements (Lessons, Left Nav, Assignments, Rich text editor).

### CODE REVIEW

**Comment by jonespm** [comment]
Looks really nice! 
A conversion script PR should be added when this db model is done to https://github.com/sakaiproject/sakai-reference/blob/master/docs/conversion/sakai_25_mysql_conversion.sql. It looks like it's already checked so that's good as a reminder.
Fix up and remove the TODO's in here. 

Otherwise looks like a nice fix and looking forward to trying it out.

**Comment by zhuoY121** [comment]
Hey Matt, thank you for the feedback! Code clean-up completed :)

> Looks really nice! A conversion script PR should be added when this db model is done to https://github.com/sakaiproject/sakai-reference/blob/master/docs/conversion/sakai_25_mysql_conversion.sql. It looks like it's already checked so that's good as a reminder. Fix up and remove the TODO's in here.
> 
> Otherwise looks like a nice fix and looking forward to trying it out.

File: basiclti/basiclti-blis/src/java/org/sakaiproject/lti/entityprovider/LTIEntityProvider.java [hunk 1]
```diff
@@ -124,6 +124,46 @@ public Map<String,Object> handleContent(EntityView view) {
                 return content;
         }
 
+	/**
+	 * Retrieve a toolSite from a siteId
+	 * /lti/toolSite/{siteId}/{toolSiteId}
+	 */
```
Comment (ern) [inline basiclti/basiclti-blis/src/java/org/sakaiproject/lti/entityprovider/LTIEntityProvider.java:130]: Add @param

File: basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/DBLTIService.java [hunk 2]
```diff
@@ -657,13 +669,21 @@ public List<Map<String, Object>> getThingsDao(String table, String[] model,
...
+				String deployWhere = "";
+				if ( includeLaunchable ) {
+					deployWhere = " OR (" + table+'.'+LTI_ID+ " IN (SELECT tool_id FROM lti_tool_site WHERE SITE_ID = ?) )";
+				}
+
 				whereClause = " ("+table+'.'+LTI_SITE_ID+" = ? OR "+
-					"("+table+'.'+LTI_SITE_ID+" IS NULL AND "+table+'.'+LTI_VISIBLE+" != 1 ) ) ";
+						"("+table+'.'+LTI_SITE_ID+" IS NULL AND "+table+'.'+LTI_VISIBLE+" != 1 ) "+deployWhere+" ) ";
 				fields.add(siteId);
+				if ( deployWhere.length() > 0 ) fields.add(siteId);
```
Comment (ern) [inline basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/DBLTIService.java:683]:
```suggestion
				if ( deployWhere.isEmpty() ) fields.add(siteId);
```

File: basiclti/basiclti-tool/src/webapp/vm/lti_tool_site_deploy.vm [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+${includeLatestJQuery}
+
+<script type="text/javascript" src="/library/webjars/jquery-ui/1.12.1/jquery-ui.min.js"></script>
+<link type="text/css" href="/library/webjars/jquery-ui/1.12.1/jquery-ui.min.css" rel="stylesheet" media="screen" />
+
+<!-- Tablesorter: required -->
+<script type="text/javascript">includeWebjarLibrary('jquery.tablesorter/2.27.7/dist/css/theme.blue.min.css');</script>
+<script type="text/javascript">includeWebjarLibrary('jquery.tablesorter');</script>
+
+<script id="js">$(function(){
```
Comment (ern) [inline basiclti/basiclti-tool/src/webapp/vm/lti_tool_site_deploy.vm:13]: this should be refactored into a js file

File: basiclti/basiclti-blis/src/bundle/lti.properties [hunk 4]
```diff
@@ -6,4 +6,5 @@ lti.action.contents=Retrieve the content items for a siteId. eg /lti/{siteId}/co
 lti.action.content=Retrieve a content item from a siteId. eg /lti/content/{siteId}/{contentId}[.format]
 lti.action.deploys=Retrieve the deployments for a siteId. eg /lti/{siteId}/deploys[.format][?first=0&last=10&order=title] (only works for !admin)
 lti.action.deploy=Retrieve a deployment from a siteId. eg /lti/deploy/{siteId}/{deployId}[.format] (only works for !admin)
-
+lti.action.toolSite=Retrieve a toolSite from a siteId. eg /lti/toolSite/{siteId}/{toolSiteId}[.format]
```
Comment (jesusmmp) [inline basiclti/basiclti-blis/src/bundle/lti.properties:9]: Are these properties used?

**Pull request merged**